### PR TITLE
Turbo Module supports intersection type for TypeScript

### PR DESCRIPTION
**Pull request by ZihanChen-MSFT**
## Summary

As [requested from community](https://github.com/reactwg/react-native-new-architecture/discussions/91#discussioncomment-4282384), intersection type is supported for component but no in turbo module. This PR fixed it.

## Changelog

[GENERAL] [CHANGED] - Turbo Module supports intersection type for TypeScript

## Test Plan

`yarn jest react-native-codegen` passed

### CODE REVIEW

File: packages/react-native-codegen/src/parsers/typescript/modules/index.js [hunk 1]
```diff
@@ -73,6 +78,64 @@ const {
...
+          cxxOnly,
+          nullable,
+          translateTypeAnnotation,
+          parser,
+        );
+      });
+    })
+    .filter(Boolean);
+
+  if (baseTypes.length === 0) {
```
Comment (cipolleschi) [inline packages/react-native-codegen/src/parsers/typescript/modules/index.js:115]: can we simplify and avoid some code duplication with:

```ts
let objectTypeAnnotation;
if (baseTypes.length === 0) {
  objectTypeAnnotation = {
        type: 'ObjectTypeAnnotation',
        properties,
      }
} else {
  objectTypeAnnotation = {
        type: 'ObjectTypeAnnotation',
        properties,
        baseTypes,
      }
}

return typeAliasResolution(
      typeAliasResolutionStatus,
      objectTypeAnnotation,
      aliasMap,
      nullable,
);
```

also, if we are using this `if length === 0 then simpleObjectTypeAnnotation else objectTypeAnnotationWithBaseType` logic in more places, perhaps we can add a factory method for it. Something like

```ts
function createObjectTypeAnnotation(properties, baseTypes) {
if (baseTypes.length === 0) {
  return {
        type: 'ObjectTypeAnnotation',
        properties,
      };
} else {
  return {
        type: 'ObjectTypeAnnotation',
        properties,
        baseTypes,
      };
}
```

and then use it like:

```ts
return typeAliasResolution(
      typeAliasResolutionStatus,
      createObjectTypeAnnotation(properties, baseTypes),
      aliasMap,
      nullable,
    );
```

What do you think?

**Review by cipolleschi — COMMENTED** [review]
Amazing work here, thank you for adding the support for these new features.

I left a comment with a couple of suggestions, but let me know what do you think!

File: packages/react-native-codegen/src/parsers/typescript/modules/index.js (hunk 1 above)
Comment (ZihanChen-MSFT) [reply packages/react-native-codegen/src/parsers/typescript/modules/index.js:116]: Fixed. I don't think we are going to repeat this logic else where so I didn't extract the function.

**Review by cipolleschi — APPROVED** [review]
looks good to me!

**Pull request closed without merging**